R/chooseFromList.R

```diff
@@ -0,0 +1,106 @@
+    message("group must have same dimension as theList, or multiple not allowed. Group mode disabled")
+    group <- FALSE
+  }
+  m <- paste0("\n\nPlease choose ", type, ":\n\n")
+  # in group mode, paste group after each entry and add them to theList as options
+  if (! isFALSE(group)) {
+    groups <- sort(unique(group))
+    groupsids <- seq(length(originalList) + 2, length(originalList) + length(groups) + 1)
+    theList <- c(paste0(str_pad(theList, max(nchar(originalList), 6), side = "right"), " ", group),
+                 paste("Group:", groups))
+    m <- c(m, paste0(str_pad("", max(nchar(originalList), 6) + nchar(length(theList) + 2) + 2, side = "right"),
+              " Group"))
+  }
+  # add all and regex pattern as options
+  if (multiple) theList <- c("all", theList, "Search by pattern...")
+  m <- c(m, paste(paste(str_pad(seq_along(theList), nchar(length(theList)), side = "left"), theList, sep = ": "),
+            collapse = "\n"))
+  m <- c(m, paste0("\nNumber", if (multiple) "s entered as 2,4:6,9",
+            if (allowEmpty) " or leave empty", " (", type, "): "))
+  if (isFALSE(userinput)) { # print options and ask for userinput
+    message(m)
+    userinput <- getLine()
+  }
+  # split userinput and perform basic checks
+  identifier <- strsplit(userinput, ",")[[1]]
+  if (allowEmpty && length(identifier) == 0) return(if (returnBoolean) booleanList else NA)
+  if (length(identifier) == 0 || ! all(grepl("^[0-9,:]*$", identifier))) {
+    message("Try again, you have to choose some numbers.")
+    return(chooseFromList(originalList, type, returnBoolean, multiple, allowEmpty, group))
+  }
+  # go through list of userinput, and turn 2:5 into 2,3,4,5 etc.
+  tmp <- NULL
+  for (i in seq_along(identifier)) {
+    if (isTRUE(length(strsplit(identifier, ":")[[i]]) > 1)) {
+      tmp <- c(tmp, as.numeric(strsplit(identifier, ":")[[i]])[1]:as.numeric(strsplit(identifier, ":")[[i]])[2])
+    } else {
+      tmp <- c(tmp, as.numeric(identifier[i]))
+    }
+  }
+  identifier <- tmp
+  # check whether all input is usable
+  if (! multiple && isTRUE(length(identifier) > 1)) {
+    message("Try again, not in list or multiple chosen: ", paste(identifier, collapse = ", "))
+    return(chooseFromList(originalList, type, returnBoolean, multiple, allowEmpty, group))
+  }
+  if (any(! identifier %in% seq_along(theList))) {
+    message("Try again, not all in list: ", paste(identifier, collapse = ", "))
+    return(chooseFromList(originalList, type, returnBoolean, multiple, allowEmpty, group))
+  }
+  # interpret group inputs
+  if (! isFALSE(group)) {
+    selectedGroups <- sub("^Group: ", "", theList[intersect(identifier, groupsids)])
+    identifier <- unique(c(identifier[! identifier %in% groupsids], which(group %in% selectedGroups) + 1))
+  }
+  # if search by pattern is selected, ask for pattern and interpret it
+  if (multiple && length(identifier == 1) && identifier == length(theList)) {
```

Comment:
here, `identifier` can be a list with more than one entry. therefore `identifier == length(theList)` can also be a list with mor than one entry. Coercing it to boolean will then raise a warning.

Reply:
```suggestion
  if (multiple && length(identifier) == 1 && identifier == length(theList)) {
```
Is probably more correct?

Reply:
Baaah, Must be `length(identifier) == 1` of course :(

Reply:
I got to the right line with `options(warn=2, error=recover)` - seems to be the go-to debugging tool for when you need to know where the warning actually happened.